contractor/processing_chain.cpp

```diff
@@ -130,81 +100,40 @@ int Prepare::Run()
-    {
-        BOOST_ASSERT_MSG(static_cast<unsigned int>(std::max(edge.weight, 1)) > 0,
-                         "edge distance < 1");
-        if (edge.forward)
-        {
-            edges.push_back({edge.source, edge.target, {}});
-        }
-
-        if (edge.backward)
-        {
-            edges.push_back({edge.target, edge.source, {}});
-        }
-    }
-
-    // connect forward and backward nodes of each edge
-    for (const auto &node : input_nodes)
-    {
-        if (node.reverse_edge_based_node_id != SPECIAL_NODEID)
-        {
-            edges.push_back({node.forward_edge_based_node_id, node.reverse_edge_based_node_id, {}});
-            edges.push_back({node.reverse_edge_based_node_id, node.forward_edge_based_node_id, {}});
-        }
-    }
-
-    tbb::parallel_sort(edges.begin(), edges.end());
-    auto new_end = std::unique(edges.begin(), edges.end());
-    edges.resize(new_end - edges.begin());
-
-    auto uncontractor_graph = std::make_shared<UncontractedGraph>(max_edge_id + 1, edges);
-
-    TarjanSCC<UncontractedGraph> component_search(
-        std::const_pointer_cast<const UncontractedGraph>(uncontractor_graph));
-    component_search.run();
-
-    for (auto &node : input_nodes)
-    {
-        auto forward_component = component_search.get_component_id(node.forward_edge_based_node_id);
-        BOOST_ASSERT(node.reverse_edge_based_node_id == SPECIAL_EDGEID ||
-                     forward_component ==
-                         component_search.get_component_id(node.reverse_edge_based_node_id));
-
-        const unsigned component_size = component_search.get_component_size(forward_component);
-        const bool is_tiny_component = component_size < 1000;
-        node.component_id = is_tiny_component ? (1 + forward_component) : 0;
+    SimpleLogger().Write() << "Opening " << edge_based_graph_filename;
+    boost::filesystem::ifstream input_stream(edge_based_graph_filename, std::ios::in | std::ios::binary);
+
+    const FingerPrint fingerprint_valid = FingerPrint::GetValid();
+    FingerPrint fingerprint_loaded;
+    input_stream.read((char *)&fingerprint_loaded, sizeof(FingerPrint));
+    fingerprint_loaded.TestPrepare(fingerprint_valid);
+
+    unsigned number_of_edges = 0;
+    size_t max_edge_id = SPECIAL_EDGEID;
+    input_stream.read((char *)&number_of_edges, sizeof(unsigned));
+    input_stream.read((char *)&max_edge_id, sizeof(size_t));
+    input_stream.read((char *)&edges_crc32, sizeof(unsigned));
+
+    edge_based_edge_list.resize(number_of_edges);
+    SimpleLogger().Write() << "Reading " << number_of_edges << " edges from the edge based graph";
+
+    // TODO: can we read this in bulk?  DeallocatingVector isn't necessarily
+    // all stored contiguously
+    for (;number_of_edges > 0; --number_of_edges) {
+        EdgeBasedEdge inbuffer;
+        input_stream.read((char *) &inbuffer, sizeof(EdgeBasedEdge));
+        edge_based_edge_list.emplace_back(std::move(inbuffer));
```

Comment:
Easiest way to speed this up is to add a method to `DeallocationVector` that return a reference to a bucket and directly read into this bucket.